### PR TITLE
Feature: Group Chats

### PR DESCRIPTION
**Pull request by Zintixx**
- Implemented Group Chats
- Added a packet handler and packet for handling chat stickers in group chats.

The most underused lame feature in the game.
Please enjoy.

P.S. This is my first time using arrays in action. So pls let me know if there's any better ways to handle things. I'm hoping I can carry over what I learn to Clubs

### CODE REVIEW

File: MapleServer2/Tools/GroupChatManager.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        public void RemoveGroupChat(GroupChat groupChat)
+        {
+            GroupChatList.Remove(groupChat.Id);
+        }
+
+        public GroupChat GetGroupChatById(int id)
+        {
+            return GroupChatList.TryGetValue(id, out GroupChat foundGroupChat) ? foundGroupChat : null;
+        }
+
```
Comment (AlanMorel) [inline MapleServer2/Tools/GroupChatManager.cs:30]: Empty blank line

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/GroupChatHandler.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            }
+
+            int groupChatCheck = 0;
+            for (int i = 0; i < 3; i++)
+            {
+                if (other.GroupChatId[i] != 0)
+                {
+                    groupChatCheck++;
+                }
+            }
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/GroupChatHandler.cs:84]:
```suggestion
            int count = other.GroupChatId.Count(x => x != 0);
```

**Review by angelotadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/GroupChat.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            Members.Add(player);
+
+            for (int i = 0; i < 3; i++)
+            {
+                if (player.GroupChatId[i] == 0)
+                {
+                    player.GroupChatId[i] = Id;
+                    break;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (sgessa) [reply MapleServer2/Types/GroupChat.cs:34]: Change it if possible, it improves readability

File: MapleServer2/PacketHandlers/Game/GroupChatHandler.cs [hunk 4]
```diff
@@ -75,13 +75,9 @@ private static void HandleInvite(GameSession session, PacketReader packet)
...
-            for (int i = 0; i < 3; i++)
-            {
-                if (other.GroupChatId[i] != 0)
-                {
-                    groupChatCheck++;
-                }
-            }
+            int count = other.GroupChatId.Count(x => x != 0);
+            
+            if (count == 3)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/GroupChatHandler.cs:80]: 🤔

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/GroupChatHandler.cs (hunk 4 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/GroupChatHandler.cs:80]: 😕 I just merged it with this fix that was suggested, but yeah it's throwing an error. I'm not too familiar with this new approach so I don't know what would be the fix for it or if there's a better fix than what I originally had.

**Review by AlanMorel — APPROVED** [review]

**Review by angelotadeucci — APPROVED** [review]

**Pull request merged**